### ISSUE TITLE
Primary instance deregisters from DBUS before entering main loop in clutter_main with Glib 2.47.4+

### ISSUE THREAD

**Issue by fossfreedom** [gmc-holle/xfdashboard#116]
I've created an autostart file /etc/xdg/autostart which runs xfdashboard in daemon mode

     /usr/bin/xfdashboard -d

Then I created a keyboard shortcut CTRL + TAB which runs /usr/bin/xfdashboard

Dashboard started.

I then pressed CTRL + TAB again - a second instance ran

from a TTY I can see the following

![pic](http://i.imgur.com/cY4miSA.png)

Pressing the second CTRL + TAB I see this:

![pic](http://i.imgur.com/3D2gD1P.png)

**Comment by fossfreedom** [comment]
desktop file

[Desktop Entry]
Name=Budgie dashboard
Comment=xfdashboard daemon
Icon=
Exec=/usr/bin/xfdashboard -d
TryExec=/usr/bin/xfdashboard
Terminal=false
Type=Application
NoDisplay=true
OnlyShowIn=Budgie;

**Comment by gmc-holle** [comment]
An autostart desktop file for xfdashboard should already be there - look for /etc/xdg/autostart/xfdashboard-autostart.desktop. Besides that your autostart desktop file looks like the original one except that yours is enabled by default.

When starting the xfdashboard with CTRL+TAB it does not recognize the daemon running in background. That's why xfdashboard is showing itself in windows view.

Could you check your ~/.xsession-errors file for warnings of xfdashboard? Could you try to start the daemon in a terminal window and then testing if CTRL+TAB is working better?

**Comment by fossfreedom** [comment]
ok - have removed my desktop file. logged out and logged in.  Confirmed there was a xfdashboard.desktop file in /etc/xdg/autostart

No daemon was seen running on login.

CTRL+TAB - dashboard started.
CTRL+TAB - dashboard within dashboard was seen.
ESCAPE followed by ESCAPE to close both dashboards.

No errors in .xsession-errors file

**Comment by fossfreedom** [comment]
with /usr/bin/xfdashboard -d running in a terminal

same thing happened - invoking via a keyboard shortcut launched xfdashboard.  Second keyboard shortcut and xfdashboard within xfdashboard was seen.

**Comment by gmc-holle** [comment]
Whatever I do, I cannot reproduce it. Sorry.

So I can only guess what may go wrong. Could you please do the following?

1.) Kill any running xfdashboard process
2.) Start xfdashboard in daemon mode in a terminal: xfdashboard -d
3.) Check that a process with ID 'de.froevel.nomad.xfdashboard' is registered on DBUS session bus. You can check it with the application d-feet or with the following command:

```
dbus-send --session --type=method_call --print-reply --dest=de.froevel.nomad.xfdashboard /de/froevel/nomad/xfdashboard org.freedesktop.DBus.Introspectable.Introspect
```

When using the command above you should get a lenghty response like:

```
method return sender=:1.170 -> dest=:1.184 reply_serial=2
   string "<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
                      "http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
<!-- GDBus 2.46.2 -->
<node>
  <interface name="org.freedesktop.DBus.Properties">
    <method name="Get">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="out"/>
    </method>
    <method name="GetAll">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="a{sv}" name="properties" direction="out"/>
    </method>
    <method name="Set">
      <arg type="s" name="interface_name" direction="in"/>
      <arg type="s" name="property_name" direction="in"/>
      <arg type="v" name="value" direction="in"/>
    </method>
    <signal name="PropertiesChanged">
      <arg type="s" name="interface_name"/>
      <arg type="a{sv}" name="changed_properties"/>
      <arg type="as" name="invalidated_properties"/>
    </signal>
  </interface>
  [...]
```

**Comment by fossfreedom** [comment]
Hi,
  xfdashboard doesnt look like it has registered itself with dbus :(

If you have virtualbox or similar I've got the budge-remix ISO uploaded - you can run xfdashboard in a liveCD (or install it as per the README).

- https://sourceforge.net/projects/budgie-remix/files/tech-demo2/

Alternatively - if you've got ubuntu 16.04 in a virtualbox session - you could add my ppa 

    sudo add-apt-repository ppa:fossfreedom/testbudgie
    sudo apt-get update
    sudo apt-get install budgie-desktop-environment xfdashboard

**Comment by fossfreedom** [comment]
Hi - I've installed xubuntu-desktop in my 16.04 VM and tried xfdashboard as well.  It too has the same issue - dbus doesnt appear to show xfdashboard registering.

What development environment are you using?  I'm wondering if there is a systemd-dbus issue here - new distro's like arch and fedora/ubuntu 16.04 are now using systemd.

As I said - speculation - I dont really know why xfdashboard isnt registering in 16.04

**Comment by fossfreedom** [comment]
@gmc-holle - just to say - I would really really really would like to ship my new distro budgie-remix (aka Ubuntu budgie) with xfdashboard.  This issue is really the only one stopping me doing the integration work.  If you can find any time whatsoever I would appreciate it - I'm not adverse to adding any debugging statements etc to help diagnose the issue.  Just let me know what & how you want me to-do.

Many thanks

**Comment by gmc-holle** [comment]
@fossfreedom You're not forgotten ;) I have tried some distributions last weekend to see if this problem is specific to a distribution or to something else. I've created a VM with budgie, ubuntu 16.04, fedora 23 and fedora 24. At all distribution (except Fedora 23) the same problem occured - xfdashboard does not register at D-Bus only Fedora 23 and Gentoo (my favourite distribution) are working correctly.

All these distribution have in common that they use Glib version 2.47 and above. The stable version at Fedora 23 and Gentoo are currently using 2.46. So something at Glib must have changed. I will now check what change could cause this issue. I ask for patience. Maybe you know someone who could help or knows this issue with Glib 2.47.x?

Believe me that I would be very happy to see xfdashboard in action at budgie :)

**Comment by fossfreedom** [comment]
this? https://lists.freedesktop.org/archives/dbus/2016-January/016846.html

Looks like glib 2.47 has removed support for dbus-glib which from the configure file xfdashboard has a dependency on.

**Comment by gmc-holle** [comment]
It's definitely Glib 2.47 (I've tested it with 2.47.5 as installed by Ubuntu 16.04 or Fedora 24). Since upgrading it on my Gentoo system it also has this problem but I don't know why :( I've not any clues.

Xfdashboard depends on the object class GApplication which is provided by the library GIO and part of the Glib library. Glib (and therefore GIO) does not depend on dbus-glib in any way so this package should have no effect on this issue. xfconf depdends on dbus-glib and that is why it is needed by xfdashboard.

I've search the web for this issue with Glib 2.47 but haven't found anything related and it seems I do anything wrong because other application depending on GApplication (or GtkApplication) do work and register on dbus.

I think I really need help here to resolve it.

**Comment by gmc-holle** [comment]
I've tested all version Glib from 2.47.1 up to latest 2.47.6. **Only version 2.47.5** is causing this problem and is the version Ubuntu 16.04 and Fedora 24 is installing. So the problem should go as soon as you upgrade Glib to version 2.47.6

**Comment by fossfreedom** [comment]
ok - thanks for the feedback.  

I'll keep an eye on ubuntu - since its in feature freeze its very very unlikely to move beyond v2.47.5 for the next 5 years which means there is very little I can do about it.  

I'll file a bug-report - but that's as much as I can do.

Many thanks for your efforts.  Well done for finding out the issue :)

**Comment by gmc-holle** [comment]
As 2.47.5 is a developement version I'm pretty sure that all distribution will upgrade to 2.48.0 at least.

Marking this issue as invalid.

**Comment by fossfreedom** [comment]
ubuntu is now at 2.47.6 of GLib - unfortunately for me xfdashboard doesn't do its nice thing.

 - http://packages.ubuntu.com/search?keywords=libglib&searchon=names&suite=wily&section=all

Looks to be exactly at the same version as upstream:

 - https://github.com/GNOME/glib/releases

I'll keep monitoring the situation and let you know if things change.

**Comment by fossfreedom** [comment]
gah - wrong link

http://packages.ubuntu.com/search?suite=xenial&searchon=names&keywords=libglib

**Comment by fossfreedom** [comment]
the only relevant dbus change during 2.47.6 series was this

 - https://github.com/GNOME/glib/commit/3c0cddfe80990a653ff7b38e1701d3950df642b6

Is this relevant to xfdashboard?

**Comment by gmc-holle** [comment]
I think I found the cause of this problem. I believe the change at commit GNOME/glib@21b1c390a3ce1f7e2816c6309f161c4b92470c46 ensures the deregistration of an application from D-BUS (the line `g_application_impl_destroy (application->priv->impl);`)

It doesn't explain why it works on my Gentoo developement system with Glib 2.47.6 but it shouldn't. Maybe this system is lightning fast ;)

Now the very sad thing. I've to rewrite my application class XfdashboardApplication and the main function from scratch - more or less. A task I'm really afraid of and I guess it will take a lot of time and will not be done anytime soon :'(

**Comment by fossfreedom** [comment]
congrats of sorts! - at least there is light at the end of the tunnel - just depends how long that tunnel is.

I'm going to do the prep work anyway to incorporate xfdashboard into budgie-remix - ideally I would like to get this into beta 1 which is end-of-march.  end of april which is the LTS date release of budgie-remix.  Sept is beta 1 for 16.10 - end of oct is release date for 16.10.  

If you have anything ready for testing at any of those dates please shout - I'm happy to do any testing you wish me to.  Thanks.

**Comment by fossfreedom** [comment]
Hi,

  having a keybinding  to launch xfdashboard works.

Using the keybinding again does not launch xfdashboard again  -  wonderful!

close via Escape

running xfdashboard -d &

then using the keybinding - xfdashboard launches - press escape to close xfdashboard

Then using the keybinding again - this time no xfdashboard launches :(

checking d-feet - the session-bus is showing de.froeval....

**Comment by gmc-holle** [comment]
I cannot reproduce it even in a Xubuntu 16.04 VM. It works as on my development system. I'm really happy with the result.

But besides that please try to following. Instead of calling just "xfdashboard" without any parameter on your key binding add paramter "-t" or "--toggle" (for long). This should work definitely - at least by my testings.

**Comment by fossfreedom** [comment]
Hi,
  the problem looks like because I'm not using the xfce window manager.

If in your VM you install another DE like gnome-shell you'll see the issue.

The -t option is quite neat - so thanks for sharing.  The new version of xfdashboard works quite nicely with the -t binding - so it is usable (but not a visually instantaneous as per xubuntu) as a solution as long as I dont run in daemon mode.

**Comment by gmc-holle** [comment]
Testing xfdashboard in Gnome Shell isn't useful as xfdashboard was born as a replacement of it ;) So it isn't testing to be running well at Gnome Shell or its window manager mutter - it isn't even tested to be running at all.

I will retry it on Fedora and maybe one or two more distribution as soon as get more spare time. In the meantime I keep the status "BLOCKER" at this issue.

**Comment by fossfreedom** [comment]
yep - makes sense.

If you've got the bandwidth you could try budgie-remix - if you do install this, make sure you do a full update.

You'll then see (after a reboot) how I've integrated xfdashboard into the DE (press CTRL+TAB to display xfdashboard)

**Comment by gmc-holle** [comment]
As promised I tried today the latest git on budgie remix. Budgie remix as installed in a virtual environment exactly the same way as suggested in the instruction plus all updates installed available today and .... xfdashboard was working as expected. I was running a daemon and in a terminal I started a loop with 5 seconds delay which called xfdashboard to trigger the daemon to activate and as soon as xfdashboard appeared fullscreen it was suspended by pressing ESCape to get it reappear 5 seconds later. All the time the daemon was kept registered at DBUS session bus.

It also worked on Xubuntu 16.04 and Fedora 24 (pre-releases).

I'm sorry and I'm glad to say: I cannot reproduce the problem anymore. So I remove the "BLOCKER" label now.
